### PR TITLE
[ini] Update ini parsing and format

### PR DESCRIPTION
**Pull request by kparichay**
Update the parsing and format of ini input file

Related issue : #239 

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: test/unittest/unittest_nntrainer_internal.cpp [hunk 1]
```diff
@@ -153,7 +152,7 @@ TEST(nntrainer_NeuralNetwork, init_07_p) {
   status = NN.setConfig("./test.ini");
   EXPECT_EQ(status, ML_ERROR_NONE);
   status = NN.init();
-  EXPECT_EQ(status, ML_ERROR_NONE);
+  EXPECT_EQ(status, ML_ERROR_INVALID_PARAMETER);
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_internal.cpp:155]: @jijoongmoon please verify is this should fail or pass. Passing no `TrainData` is acceptable?

Comment (kparichay) [reply test/unittest/unittest_nntrainer_internal.cpp:155]: Can you also clarify on `LabelData`, it's used by all train/valid/test it seems, That's a bit odd.

**Comment by zhoonit** [comment]
Could you elaborate As-is / To-be of this PR if possible?
Thanks!

File: nntrainer/src/neuralnet.cpp [hunk 2]
```diff
@@ -124,118 +124,137 @@ int NeuralNetwork::setConfig(std::string config) {
...
-    NN_INI_RETURN_STATUS();
-
-  } else if (iniparser_find_entry(ini, "DataSet:Tflite")) {
-    ml_loge("Error: Not yet implemented!");
-    return ML_ERROR_INVALID_PARAMETER;
-  } else {
-    data_buffer = std::make_shared<DataBufferFromCallback>();
+  /** Parse the DataSet section */
+  data_buffer = std::make_shared<DataBufferFromCallback>();
+  section_names_iter =
```
Comment (jijoongmoon) [inline nntrainer/src/neuralnet.cpp:204]: This 203 line has conflict with line 212

File: nntrainer/src/neuralnet.cpp [hunk 3]
```diff
@@ -124,118 +124,137 @@ int NeuralNetwork::setConfig(std::string config) {
...
+      NN_INI_RETURN_STATUS();
+    }
   }
 
+  /** Parse all the layers defined as sections in order */
   TensorDim previous_dim;
-  for (unsigned int i = 0; i < layers_name.size(); i++) {
+  for (section_names_iter = section_names.begin();
+      section_names_iter != section_names.end(); ++section_names_iter) {
     bool last = false;
```
Comment (jijoongmoon) [inline nntrainer/src/neuralnet.cpp:235]: So, first layer should go first and last layer should go last, then. Strictly follow in order.

File: nntrainer/src/neuralnet.cpp [hunk 4]
```diff
@@ -948,44 +1007,43 @@ int NeuralNetwork::addLayer(std::shared_ptr<Layer> layer) {
...
-  ActiType act = layer->getActivationType();
-  if (act != ACT_UNKNOWN) {
-    if (layers.back()->getType() == LAYER_ACTIVATION) {
-      /* possible throw here? */
-      ml_logw("Activation defined right after activation");
-    }
+std::shared_ptr<Layer> NeuralNetwork::_make_act_layer(ActiType act) {
+  if (layers.back()->getType() == LAYER_ACTIVATION) {
+    /** User defined activation layer. Do not add another activation layer after this */
+    ml_loge("Error: double activation layers.");
```
Comment (jijoongmoon) [inline nntrainer/src/neuralnet.cpp:1013]: Could you talk to @zhoonit about it? He has intention to have warning instead of error.

File: test/unittest/unittest_nntrainer_internal.cpp (hunk 1 above)
Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_internal.cpp:155]: 1. If you mean the TrainData in ini file,  it's acceptable. We have two options to get training data. one is from file. and the other is from callback which defined by user.
2. Label Data keyword is used to calculate loss. Is there any other way to get the label data?

Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_internal.cpp:155]: Discussed off-line and make agreement.

File: nntrainer/src/neuralnet.cpp (hunk 4 above)
Comment (zhoonit) [reply nntrainer/src/neuralnet.cpp:1013]: NBD, I think this can easily be fixed it it becomes a requirement :)

File: nntrainer/include/layer.h [hunk 5]
```diff
@@ -123,7 +123,7 @@ class Layer {
       type(LAYER_UNKNOWN),
       loss(0.0),
       cost(COST_UNKNOWN),
-      activation_type(ACT_UNKNOWN),
+      activation_type(ACT_NONE),
```
Comment (zhoonit) [inline nntrainer/include/layer.h:126]: Please note that, ACT_NONE will set an layer with identity activation function w.r.t. current implementation of `activation layer` in `neuralnet::_make_act_layer`

File: nntrainer/src/neuralnet.cpp [hunk 6]
```diff
@@ -550,30 +585,51 @@ int NeuralNetwork::init(std::shared_ptr<Optimizer> optimizer,
   status = setProperty(arg_list);
   NN_RETURN_STATUS();
 
-  loss = 10000.0;
+  /** Note: number of entries in layers will change. */
```
Comment (zhoonit) [inline nntrainer/src/neuralnet.cpp:588]: This is good

File: nntrainer/include/layer.h (hunk 5 above)
Comment (kparichay) [reply nntrainer/include/layer.h:126]: Right, I updated to result in an error if the activation type is unknown.
IMO itis better to make it a no-op (no activation) than keep it unknown. Regarding the overhead, unit layers can be optimized later.
Can update it to `unknown` if you think its better.

File: nntrainer/src/neuralnet.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:1013]: In my understanding, this situation only happens due to internal bug as user cannot request for a separate LAYER_ACTIVATION as of now. So, I changed this to error.

File: nntrainer/src/neuralnet.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:235]: Yes, just like the older ini format required `Layers` to give the order of layers.
```
Layers = inputlayer \
         fc1layer \
         batchnormalization \
        outputlayer    #Layers of Neuralnetwork
```

File: nntrainer/src/neuralnet.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:204]: Yes, this has been updated.

File: nntrainer/include/layer.h (hunk 5 above)
Comment (zhoonit) [reply nntrainer/include/layer.h:126]: I think `ACT_NONE` is better :)

File: nntrainer/src/neuralnet.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/src/neuralnet.cpp:240]: Yes, this is the location where the order of layers is defined.
Please refer https://github.com/nnstreamer/nntrainer/pull/240#issuecomment-651039288 for detailed example.
